src/utils/BlockUtils.ts

```diff
@@ -166,3 +169,40 @@ export class BlockFinder {
     }
   }
 }
+
+/**
+ * @notice Get the block number for a given timestamp fresh from on-chain data if not found in redis cache.
+ * If redis cache is not available, then requests block from blockFinder.
+ * @param chainId Chain to load block finder for.
+ * @param timestamp Approximate timestamp of the to requested block number.
+ * @param blockFinder Caller can optionally pass in a block finder object to use instead of creating a new one
+ * or loading from cache. This is useful for testing primarily.
+ * @returns Block number for the requested timestamp.
+ */
+export async function getCachedBlockForTimestamp(
+  chainId: number,
+  timestamp: number,
+  blockFinder: BlockFinder,
+  provider: Provider,
+  cache?: CachingMechanismInterface
+): Promise<number> {
+  // If no redis client, then request block from blockFinder. Otherwise try to load from redis cache.
+  if (!isDefined(cache)) {
+    return (await blockFinder.getBlockForTimestamp(timestamp)).number;
+  }
+  const key = `${chainId}_block_number_${timestamp}`;
+  const result = await cache.get<string>(key);
+  if (result === null) {
+    const [currentBlock, { number: blockNumber }] = await Promise.all([
+      provider.getBlock("latest"),
+      blockFinder.getBlockForTimestamp(timestamp),
+    ]);
+    // Expire key after 90 days.
+    if (shouldCache(timestamp, currentBlock.timestamp, DEFAULT_CACHING_SAFE_LAG)) {
```

Comment:
Once again, I think it would be ideal to avoid extra get latest block calls if possible -- any reason we can't just use the current local machine's time?

Reply:
Resolved https://github.com/across-protocol/sdk-v2/pull/438/commits/2015da59e59943f7babf309bc60ceab5601fe315